### PR TITLE
487-Added sound on 10 seconds countdown.

### PR DESCRIPTION
**Pull request by h4rdmol**
Hello guys, 

According to the ticket https://trello.com/c/dMLSHoJH/487-countdown-timer-sound I have added a countdown timer. 
I've used TickTock sound from the link attached to the ticket. 

**The solution:**
I have checked that the timer text gets updated in Widget Blueprint, so I have used PlayTimerTickTack() function. The function is implemented in SoundsManager class.  
* UpdateInfo gets called a couple of times, depending on InGameFrequency (0.25 for my PC). So there were about 3-4 calls for PlaySound2D and the sound gets overlapped. 
* I've tried to do the calculation of the correct time to play each 1 second the audio file was having a delay in starting playing. 

**So my solution is next:** 
* hardcode seconds to start countdown timer playing to 10 seconds in BluePrint
* extend the file to 10 seconds to match.

The countdown timer works but still, I don't have a feeling it's a good solution, since the tickTock sound is now 2Mb instead of 200kB.

P.S. I am a newbie in UE development and just finished the N-iX course in Lviv. Would like to do programming as a hobby. Will be happy for feedback and review.

### CODE REVIEW

**Comment by h4rdmol** [comment]
Looks like #486 is also done as part of PR. 
https://trello.com/c/r2SMdEnZ/486-blend-out-the-main-menu-music-and-blend-in-the-level-music-during-the-countdown-timer

File: Source/Bomber/Public/GameFramework/MyGameStateBase.h [hunk 1]
```diff
@@ -75,7 +75,7 @@ class AMyGameStateBase final : public AGameStateBase
 	/** Seconds to the end of the round,
 	 * @todo Move to Data asset */
 	UPROPERTY(EditDefaultsOnly, BlueprintReadWrite, meta = (BlueprintProtected, DisplayName = "In-Game Countdown"))
-	int32 InGameCountdownInternal = 120; //[B]
+	int32 InGameCountdownInternal = 120; //[B]
```
Comment (JanSeliv) [inline Source/Bomber/Public/GameFramework/MyGameStateBase.h:78]: Please remove MyGameStateBase.h from the review, it contains almost no changes

File: Binaries/Win64/BomberEditor.target [hunk 2]
```diff
@@ -11,8 +11,8 @@
 	{
 		"MajorVersion": 5,
 		"MinorVersion": 0,
-		"PatchVersion": 2,
```
Comment (JanSeliv) [inline Binaries/Win64/BomberEditor.target:14]: Please remove all binaries (Binaries/Win64) from review, we will put binaries once with merging the Dev with Master

**Comment by JanSeliv** [comment]
I'm so sorry there were no descriptions in the trello card, the task is about to add sounds for the 3-2-1 countdown timer that is happening during 'Game Starting' state before the actual match starts
BUT your idea I also like! It looks reasonable to add sounds for last seconds of the game, so let's review this proposition

**Review by JanSeliv — CHANGES_REQUESTED** [review]
#1 Instead of hardcoding the seconds as '10' you can take the length directly from the sound itself, so if the designer wants to change the sound, then it will not break anything
#2 The equal operator there for float could cause missing the sound If some lag happens, I recommend compare it at least withing the Delta Seconds range.
#3 You don't really need to 'cache' the Sounds Manager having its own local variable. Call of 'Get Sounds Manager' costs nothing since it returns you directly the pointer without iterating anything
image

**Review by JanSeliv — CHANGES_REQUESTED** [review]
**WBP_InGame** is UI widget that represents logic to display things on UI only.
Such logic about trigger the sound X seconds before ending the game would be better put to a more responsible place like **MyGameStateBase** since it is responsible for all time in the game.

The **AMyGameStateBase::DecrementInGameCountdown()** could be a good place for it (it exists on Develop and is not merged to Master yet):
```
void AMyGameStateBase::DecrementInGameCountdown()
{
	if (CurrentGameStateInternal != ECGS::InGame)
	{
		return;
	}

	const float TickInterval = UGameStateDataAsset::Get().GetTickInterval();
	InGameTimerSecRemainInternal -= TickInterval;

	if (IsInGameTimerElapsed())
	{
		ServerSetGameState(ECurrentGameState::EndGame);
	}
	else
	{
		const USoundBase* TickTackSFX = USoundsDataAsset::Get().GetExplosionSFX();
		const USoundsManager* SoundsManager = USingletonLibrary::GetSoundsManager();
		if (TickTackSFX && SoundsManager)
		{
			const float SoundDuration = TickTackSFX->GetDuration();
			const float Tolerance = TickInterval - KINDA_SMALL_NUMBER + SoundDuration % TickInterval;
			if (FMath::IsNearlyEqual(InGameTimerSecRemainInternal, SoundDuration, Tolerance))
			{
				SoundsManager->PlayTimerTickTackSFX();
			}
		}
	}
}
```
Note: I'm not sure about 'Tolerance'. I put '+ SoundDuration % TickInterval' to match with the difference since 'Tick Interval' could not match with the length of the sound, so the condition could be skipped. '- KINDA_SMALL_NUMBER' to avoid entering more than once. Probably you need to experiment with it and check does it work.

**Comment by JanSeliv** [comment]
Because of the reason conflicts (WBP_InGame is already changed on Dev), we can't submit these changes there on master. Could you please address feedback and prepare another PR on Develop? I would really appreciate it!

**Review by JanSeliv — COMMENTED** [review]
Please use next tweaks for the SW_TickTack the same for all other Sound Waves in the game. These tweaks allow it to be played correctly and its volume level could be changed by player right in the game settings menu:
- Sound Asset Compression Type: PCM
- Class: BomberSFX
image

**Pull request closed without merging**